src/index_factory.cpp

```diff
@@ -0,0 +1,248 @@
+#include "index_factory.h"
+
+namespace diskann
+{
+
+IndexFactory::IndexFactory(const IndexConfig &config) : _config(config)
+{
+    check_config();
+}
+
+std::unique_ptr<AbstractIndex> IndexFactory::get_instance()
+{
+    size_t num_points = _config.max_points;
+    size_t dim = _config.dimension;
+    if (_config.data_type == "float")
```

Comment:
This seems like a pretty complex `if`/`else` tree. Can we simplify it by recursively swapping one string parameter into a compile parameter at a time? Something like:

```
std::unique_ptr<AbstractIndex> IndexFactory::create_instance()
{
    ... 
    return create_instance(_config.data_type, _config.tag_type, _config.label_type);
}

std::unique_ptr<AbstractIndex> create_instance<T>(const std::string& data_type, const std::string& tag_type, const std::string& label_type)
{
    if (data_type == "float")
    {
        create_instance<float>(tag_type, label_type);
    }
    else if (data_type == ...)
...


template <typename DataType>
std::unique_ptr<AbstractIndex> create_instance<T>(const std::string& tag_type, const std::string& label_type)
{
    if (tag_type == "uint64")
    {
        create_instance<DataType, uint64_t>(label_type);
    }
    else if ...
}

... same for label_type
```